shell/browser/extensions/extension_tab_util.cc

```diff
@@ -0,0 +1,265 @@
+      api_permission =
+          extension->permissions_data()->HasAPIPermission(APIPermission::kTab);
+    } else {
+      api_permission = extension->permissions_data()->HasAPIPermissionForTab(
+          tab_id, APIPermission::kTab);
+    }
+
+    bool host_permission = extension->permissions_data()
+                               ->active_permissions()
+                               .HasExplicitAccessToOrigin(url);
+    has_permission = api_permission || host_permission;
+  }
+
+  if (!has_permission) {
+    return ExtensionTabUtil::kScrubTabFully;
+  }
+
+  return ExtensionTabUtil::kDontScrubTab;
+}
+
+ExtensionTabUtil::ScrubTabBehavior ExtensionTabUtil::GetScrubTabBehavior(
+    const Extension* extension,
+    Feature::Context context,
+    content::WebContents* contents) {
+  auto* api_wc = electron::api::WebContents::From(contents);
+
+  ScrubTabBehavior behavior;
+  behavior.committed_info = GetScrubTabBehaviorImpl(
+      extension, context, contents->GetLastCommittedURL(), api_wc->ID());
+  NavigationEntry* entry = contents->GetController().GetPendingEntry();
+  GURL pending_url;
+  if (entry) {
+    pending_url = entry->GetVirtualURL();
+  }
+  behavior.pending_info =
+      GetScrubTabBehaviorImpl(extension, context, pending_url, api_wc->ID());
+  return behavior;
+}
+
+void ExtensionTabUtil::ScrubTabForExtension(
+    const Extension* extension,
+    content::WebContents* contents,
+    api::tabs::Tab* tab,
+    ExtensionTabUtil::ScrubTabBehavior scrub_tab_behavior) {
+  // Remove sensitive committed tab info if necessary.
+  switch (scrub_tab_behavior.committed_info) {
+    case kScrubTabFully:
+      tab->url.reset();
+      tab->title.reset();
+      tab->fav_icon_url.reset();
+      break;
+    case kScrubTabUrlToOrigin:
+      tab->url =
+          std::make_unique<std::string>(GURL(*tab->url).GetOrigin().spec());
+      break;
+    case kDontScrubTab:
+      break;
+  }
+
+  // Remove sensitive pending tab info if necessary.
+  if (tab->pending_url) {
+    switch (scrub_tab_behavior.pending_info) {
+      case kScrubTabFully:
+        tab->pending_url.reset();
+        break;
+      case kScrubTabUrlToOrigin:
+        tab->pending_url = std::make_unique<std::string>(
+            GURL(*tab->pending_url).GetOrigin().spec());
+        break;
+      case kDontScrubTab:
+        break;
+    }
+  }
+}
+
+bool ExtensionTabUtil::GetTabById(
+    int tab_id,
+    content::BrowserContext* context,
+    bool include_incognito,
+    electron::api::WebContents** contents,
+    std::unique_ptr<electron::api::ChromeTabDetails>* out_tab) {
+  if (tab_id == api::tabs::TAB_ID_NONE)
+    return false;
+
+  auto* wc = electron::api::WebContents::FromID(tab_id);
+  if (!wc)
+    return false;
+
+  auto tab =
+      electron::api::Session::FromBrowserContext(context)->GetChromeTabDetails(
+          wc);
+  if (!tab)
+    return false;
+
+  if (out_tab)
+    *out_tab = std::move(tab);
+  if (contents)
+    *contents = wc;
```

Comment:
It's a bit weird that this function returns two separate things. Can we separate this out into "GetWebContentsById" and "GetTabDetailsForWebContents"?

Reply:
Sure